cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/TopicConfigurationParameters.java

```diff
@@ -17,6 +20,14 @@
  *
  * <ul>
  *   <li>Note that "review_id" is mutually exclusive to the other parameters -- i.e. they cannot be used together.</li>
+ *   <li>If topics to change replication factor and target replication factor is not specified in URL, they can also be
```

Comment:
1. I am worried that this format makes the `topic_configuration` endpoint almost dedicated to replication factor change requests for topics. Is it possible to preserve the generic purpose of this endpoint by adding a relevant `key` to the provided configuration body? -- i.e. can users specify what topic configuration to change first in a `key`, then indicate the relevant configs under that `key`?

2. I think ideally each JSON value should have an associated key that describes what the value is -- i.e. in the current design, it is not clear what the following file indicates (but we know that it is a map of topic patterns by replication factor):
```
{
    3: E.*,
    2: K.*
}
```
Note that with such an explicit naming, users can also specify parameters that this endpoint requires as input, such as `goals`, under the relevant config change request type within the JSON payload.

Reply:
Would something like the following be sufficiently well-scoped?
```
{
    "replication_factor": [
        {"3": "E.*"},
        {"2": "K.*"}
    ]
}
```
Full JSON payload support for all potentially-passed parameters, even only for the `topic_configuration` endpoint, seems like it is a larger issue, even if it is a valid method of sending information to `cruise-control`'s API.

Reply:
@efeg  for 1, I feel it might be unnecessary for this patch. What we want to achieve is that if certain parameter is too long, we specify it in request body. So ideally it should be 1:1 mapping with whatever valid parameter for this endpoint. I think it is a good idea that we preserve what parameter is for what purpose(as described in issue #886 ), but explicitly have another layer in request body seems too much

for 2, I agree, I will add a field name `topic_pattern_by_replication_factor` to accurately describe what the value represent.

Reply:
If we intend to postpone the work to support for updating general topic configurations to https://github.com/linkedin/cruise-control/issues/886, I think we should still strive to make sure that this patch does not introduce conflicting/breaking changes for the intended work.

> So ideally it should be 1:1 mapping with whatever valid parameter for this endpoint.

Achieving a 1-1 mapping may not be possible for several reasons. For example, consider the case where the user intends to change `config-1` of `selected-topics-1` and `config-2` of `selected-topics-2`, where `selected-topics-1` != `selected-topics-2`; hence, `topic` parameter wouldn't be common for these changes.

Reply:
agree,  since we want to support specifying multiple pair of replication factor and topic pattern, we are unable to do 1-1 mapping for `topic` and `replication_factor` parameter. This patch allows client to specify the pairs in body under `topic_by_replication_factor`. the format is 

>    {
      topic_by_replication_factor : {
          target_replication_factor_1 : topic_regex_1,
          target_replication_factor_2 : topic_regex_2,
          ...
      }
    }

Reply:
I am concerned that if we avoid adding a top-layer `request-type` (e.g. `replication_factor`) in this patch, any followup patch (e.g. https://github.com/linkedin/cruise-control/issues/886) will need to introduce backwards incompatible changes in the payload structure.

Here's what I propose:

```
{
    replication_factor: {
        topic_by_replication_factor : {
            target_replication_factor_1 : topic_regex_1
            target_replication_factor_2 : topic_regex_2,
            ...
    }
}
```

Reply:
Had an offline discussion, we want to support specifying multiple different configuration changes for different topics via the same POST body. 
Thus having an extra layer of `request-type` helps to organize parameter for different sub-requests in one HTTP request.

Reply:
@mgrubent FYI, the final POST body format have another layer to specify `request-type`. As Efe mentioned, since `admin` and `topic_configuration` are "multi-purpose" endpoints, another layer helps to clarify in body which parameter goes to which sub-request.

Reply:
@kidkun Got it; I think with this schema you actually allow for all types of information which are currently passed through a combination of endpoint and parameter-value mappings, to be passed through the POST body instead.

I'm not sure how far you want to allow the use of the POST body, from an API perspective, but I agree that this additional `request-type` wrapper solves the narrow problem of allowing for a sufficiently-well-scoped passing of `target_replication_factor : topic_regex` mappings

Reply:
So far, `target_replication_factor : topic_regex` is the only thing client can specify in body and that is solely because the regex could be too long.
Currently I do not have plan to parse other parameter from body, but if things changes, like we can do more and more things to these two endpoints, we may consider to support specifying all parameters in the body. Then this hierarchy in body can be beneficial.